File: cache/cache-annotation-processor/src/main/java/ru/tinkoff/kora/cache/annotation/processor/aop/AbstractAopCacheAspect.java
package ru.tinkoff.kora.cache.annotation.processor.aop;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import ru.tinkoff.kora.aop.annotation.processor.KoraAspect;
import ru.tinkoff.kora.cache.annotation.processor.CacheOperation;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import java.util.ArrayList;
import java.util.List;

import static com.squareup.javapoet.CodeBlock.joining;

abstract class AbstractAopCacheAspect implements KoraAspect {

    private static final ClassName KEY_CACHE = ClassName.get("ru.tinkoff.kora.cache", "CacheKey");

    ClassName getCacheKey(CacheOperation operation) {
        return KEY_CACHE;
    }

    List<String> getCacheFields(CacheOperation operation, ProcessingEnvironment env, AspectContext aspectContext) {
        final List<String> cacheFields = new ArrayList<>();
        for (var cacheImpl: operation.cacheImplementations()) {
            var cacheElement = env.getElementUtils().getTypeElement(cacheImpl);
            var fieldCache = aspectContext.fieldFactory().constructorParam(cacheElement.asType(), List.of());
            cacheFields.add(fieldCache);
        }

        return cacheFields;
    }

    String getKeyRecordParameters(CacheOperation operation, ExecutableElement method) {
        return String.join(", ", operation.getParametersNames(method));
    }

    String getSuperMethod(ExecutableElement method, String superCall) {
        return method.getParameters().stream()
            .map(p -> CodeBlock.of("$L", p))
            .collect(joining(", ", superCall + "(", ")")).toString();
    }
}

File: cache/cache-symbol-processor/src/main/kotlin/ru/tinkoff/kora/cache/symbol/processor/aop/AbstractAopCacheAspect.kt
package ru.tinkoff.kora.cache.symbol.processor.aop

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.AnnotationSpec
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.cache.symbol.processor.CacheOperation
import ru.tinkoff.kora.common.Tag
import java.util.*
import java.util.stream.Collectors
import javax.annotation.processing.ProcessingEnvironment

@KspExperimental
abstract class AbstractAopCacheAspect : KoraAspect {

    private val KEY_CACHE = ClassName("ru.tinkoff.kora.cache", "CacheKey")

    open fun getCacheKey(operation: CacheOperation): ClassName {
        return KEY_CACHE
    }

    open fun getCacheFields(
        operation: CacheOperation,
        resolver: Resolver,
        aspectContext: KoraAspect.AspectContext
    ): List<String> {
        val cacheFields: MutableList<String> = ArrayList()
        for (cacheImpl in operation.cacheImplementations) {
            val cacheElement = resolver.getClassDeclarationByName(cacheImpl)
            val fieldCache: String = aspectContext.fieldFactory.constructorParam(cacheElement!!.asType(listOf()), listOf())
            cacheFields.add(fieldCache)
        }
        return cacheFields
    }

    open fun getKeyRecordParameters(operation: CacheOperation, method: KSFunctionDeclaration): String {
        return operation.getParametersNames(method).joinToString(", ")
    }

    open fun getSuperMethod(method: KSFunctionDeclaration, superCall: String): String {
        return method.parameters.joinToString(", ", "$superCall(", ")")
    }
}
